### PR TITLE
Make cutout compatible with keras layer

### PR DESCRIPTION
**Pull request by WindQAQ**
# Description

Fix cutout in keras layer (graph and symbolic tensor).

## Type of change

- [x] Bug fix
- [ ] New Tutorial
- [ ] Updated or additional documentation
- [x] Additional Testing
- [ ] New Activation and the changes conform to the [activation contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/activations/README.md#contribution-guidelines)
- [ ] New Callback and the changes conform to the [callback contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/callbacks/README.md#contribution-guidelines)
- [ ] New Image addition and the changes conform to the [image op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/image/README.md#contribution-guidelines)
- [ ] New Layer and the changes conform to the [layer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/layers/README.md#contribution-guidelines)
- [ ] New Loss and the changes conform to the [loss contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/losses/README.md#contribution-guidelines)
- [ ] New Metric and the changes conform to the [metric contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/metrics/README.md#contribution-guidelines)
- [ ] New Optimizer and the changes conform to the [optimizer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/optimizers/README.md#contribution-guidelines)
- [ ] New RNN Cell and the changes conform to the [rnn contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/rnn/README.md#contribution-guidelines)
- [ ] New Seq2seq addition and the changes conform to the [seq2seq contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/seq2seq/README.md#contribution-guidelines)
- [ ] New Text addition and the changes conform to the [text op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/text/README.md#contribution-guidelines)

# Checklist:

- [x] I've properly [formatted my code according to the guidelines](https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md#coding-style)
    - [ ] By running Black + Flake8
    - [ ] By running pre-commit hooks
- [ ] This PR addresses an already submitted issue for TensorFlow Addons
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] This PR contains modifications to C++ custom-ops

# How Has This Been Tested?

Passes existing and new tests.

### CODE REVIEW

File: tensorflow_addons/image/cutout_ops.py [hunk 1]
```diff
@@ -28,8 +28,9 @@ def _norm_params(images, mask_size):
         )
     if tf.rank(mask_size) == 0:
         mask_size = tf.stack([mask_size, mask_size])
-    images_shape = tf.shape(images)
-    return mask_size, images_shape[1], images_shape[2]
+    if offset is not None and tf.rank(offset) == 1:
+        offset = tf.expand_dims(offset, 0)
+    return mask_size, offset
```
Comment (WindQAQ) [inline tensorflow_addons/image/cutout_ops.py:33]: Manage all graph-incompatible ops into this function with `tf.function`

File: tensorflow_addons/image/cutout_ops.py [hunk 2]
```diff
@@ -134,31 +147,38 @@ def cutout(
...
         cutout_shape = tf.transpose(
             [
-                image_height - (lower_pads + upper_pads),
                 image_width - (left_pads + right_pads),
+                image_height - (lower_pads + upper_pads),
             ],
             [1, 0],
         )
-        masks = tf.TensorArray(images.dtype, 0, dynamic_size=True)
-        for i in tf.range(tf.shape(cutout_shape)[0]):
```
Comment (WindQAQ) [inline tensorflow_addons/image/cutout_ops.py:143]: Iterating Tensor is not viable in graph mode, so use `tf.map_fn` instead.

File: tensorflow_addons/image/cutout_ops.py [hunk 3]
```diff
@@ -134,31 +147,38 @@ def cutout(
...
+            fn_output_signature=tf.TensorSpec(
+                shape=image_static_shape[1:-1], dtype=tf.bool
+            ),
+        )
+        mask = tf.expand_dims(mask, -1)
+        mask = tf.tile(mask, [1, 1, 1, channels])
 
         images = tf.where(
-            tf.equal(mask, 0),
-            tf.ones_like(images, dtype=images.dtype) * constant_values,
```
Comment (WindQAQ) [inline tensorflow_addons/image/cutout_ops.py:160]: `SelectV2` can handle broadcast semantics.

File: tensorflow_addons/image/tests/cutout_ops_test.py [hunk 4]
```diff
@@ -33,7 +33,7 @@ def test_different_dtypes(dtype):
 
 
 def test_different_channels():
-    for channel in [0, 1, 3, 4]:
+    for channel in [1, 2, 3, 4]:
```
Comment (WindQAQ) [inline tensorflow_addons/image/tests/cutout_ops_test.py:36]: It makes no sense to me to test 0 channel images. @fsx950223 Is this intended? Thanks.

**Comment by WindQAQ** [comment]
No performance regression on switching implementation I believe.

https://colab.research.google.com/drive/1MlPkCziMK4msWIPC-L9sJvEePqtE-rCg?usp=sharing

File: tensorflow_addons/image/cutout_ops.py [hunk 5]
```diff
@@ -139,26 +152,33 @@ def cutout(
...
-                constant_values=1,
+                constant_values=True,
             )
-            masks = masks.write(i, mask)
-
-        mask_4d = tf.expand_dims(masks.stack(), -1)
-        mask = tf.tile(mask_4d, [1, 1, 1, tf.shape(images)[-1]])
+            return mask
+
+        mask = tf.map_fn(
```
Comment (fsx950223) [inline tensorflow_addons/image/cutout_ops.py:168]: Is vectorized_map a better choice?

File: tensorflow_addons/image/tests/cutout_ops_test.py (hunk 4 above)
Comment (fsx950223) [reply tensorflow_addons/image/tests/cutout_ops_test.py:36]: It's a edge test case.

File: tensorflow_addons/image/cutout_ops.py (hunk 5 above)
Comment (WindQAQ) [reply tensorflow_addons/image/cutout_ops.py:168]: it somehow traces the graph multiple times, see the colab link above.

File: tensorflow_addons/image/tests/cutout_ops_test.py (hunk 4 above)
Comment (WindQAQ) [reply tensorflow_addons/image/tests/cutout_ops_test.py:36]: sure, let me checkout it.

File: tensorflow_addons/image/cutout_ops.py (hunk 5 above)
Comment (fsx950223) [reply tensorflow_addons/image/cutout_ops.py:168]: Yeah, map_fn is more suitable

**Review by fsx950223 — APPROVED** [review]

**Review by seanpmorgan — APPROVED** [review]
LGTM thanks! Thanks @fsx950223 for the review!
